crates/adapters/src/transport/http/input.rs

```diff
@@ -0,0 +1,411 @@
+        self.endpoint.add_socket(addr);
+    }
+
+    fn stopped(&mut self, ctx: &mut Self::Context) {
+        // Connection closed: unregister the actor.
+        let addr = ctx.address();
+        self.endpoint.remove_socket(&addr);
+    }
+}
+
+impl StreamHandler<Result<WsMessage, WsProtocolError>> for HttpInputWs {
+    /// Handle websocket message from the client.
+    fn handle(&mut self, msg: Result<WsMessage, WsProtocolError>, ctx: &mut Self::Context) {
+        match msg {
+            // The client may check our heartbeat by sending ping messages.
+            // Reply with a pong message.
+            Ok(WsMessage::Ping(msg)) => ctx.pong(&msg),
+            // Normally we send status notifications to the client whenever endpoint
+            // state changes; however the client may request a state notification, e.g.,
+            // right after connection has been established.
+            Ok(WsMessage::Text(cmd)) if cmd == "state" => {
+                self.send_state(ctx, self.endpoint.state())
+            }
+            // Data buffer received in running state.  The first 8 bytes of the buffer
+            // contain a client-assigned tag.  We acknowledge the message by sending a
+            // 16-byte response, where the first 8 bytes contain a copy of the message tag;
+            // and the rest of the message contains big-endian-encoded total message
+            // length, including the 8-byte header.
+            Ok(WsMessage::Binary(bin)) if self.endpoint.state() == PipelineState::Running => {
+                debug!(
+                    "HTTP input endpoint '{}': received {}-byte message",
+                    self.endpoint.name(),
+                    bin.len()
+                );
+                if bin.len() < 8 {
+                    return;
+                }
+                // Acknowledgement message.
+                let mut ack: [u8; 16] = [0; 16];
+                self.endpoint.push_bytes(&bin[8..]);
+                ack[0..8].copy_from_slice(&bin[0..8]);
+                (&mut ack[8..16])
+                    .write_u64::<BigEndian>(bin.len() as u64)
+                    .unwrap();
+                ctx.binary(Bytes::copy_from_slice(&ack));
+            }
+            // Data received in paused state.  Drop the message and send a negative
+            // acknowledgement, consisting of message tag followed by 8 zero bytes.
+            Ok(WsMessage::Binary(bin)) if self.endpoint.state() == PipelineState::Paused => {
+                debug!(
+                    "HTTP input endpoint '{}': rejecting {}-byte message in paused state",
+                    self.endpoint.name(),
+                    bin.len()
+                );
+
+                let mut ack: [u8; 16] = [0; 16];
+                ack[0..8].copy_from_slice(&bin[0..8]);
+                ctx.binary(Bytes::copy_from_slice(&ack));
+            }
+            // First frame of a multipart message.  Save message tag, so we can send an
```

Comment:
there isn't a library function to receive a multipart message?
When I used websockets I never worried about this stuff.
Maybe my messages weren't long enough

Reply:
I don't think the actix framework supports that today.  But the framework you used in JS most likely does.